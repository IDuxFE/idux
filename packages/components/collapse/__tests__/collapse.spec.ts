import { mount } from '@vue/test-utils'
import { Ref, ref, nextTick } from 'vue'
import { renderWork } from '@tests'
import IxCollapse from '../src/Collapse.vue'
import IxCollapsePanel from '../src/CollapsePanel.vue'

const CombineComponent = {
  components: { IxCollapse, IxCollapsePanel },
  template: `
  <ix-collapse>
    <ix-collapse-panel name="name1"> name1 </ix-collapse-panel>
    <ix-collapse-panel name="name2"> name2 </ix-collapse-panel>
    <ix-collapse-panel name="name3"> name3 </ix-collapse-panel>
  </ix-collapse>
  `,
}
describe('Collapse.vue and CollapsePanel.vue', () => {
  renderWork(CombineComponent)

  test('active(v-model) work', async () => {
    const active = ref([]) as Ref<string[]>
    const wrapper = mount({
      components: { IxCollapse, IxCollapsePanel },
      template: `
      <ix-collapse v-model:active="active">
        <ix-collapse-panel name="name1"> name1 </ix-collapse-panel>
        <ix-collapse-panel name="name2"> name2 </ix-collapse-panel>
        <ix-collapse-panel name="name3"> name3 </ix-collapse-panel>
      </ix-collapse>
      `,
      setup() {
        return { active }
      },
    })

    expect(wrapper.findAll('.ix-collapse-panel').length).toBe(3)
    expect(wrapper.findAll('.ix-collapse-panel-active').length).toBe(0)

    active.value = ['name1']

    await nextTick()

    expect(wrapper.findAllComponents({ name: 'IxCollapsePanel' })[0].classes()).toContain('ix-collapse-panel-active')
    expect(active.value).toEqual(['name1'])

    await wrapper.findAll('.ix-collapse-panel-header')[0].trigger('click')

    await nextTick()

    expect(wrapper.findAllComponents({ name: 'IxCollapsePanel' })[0].classes()).not.toContain(
      'ix-collapse-panel-active',
    )
    expect(active.value).toEqual([])

    await wrapper.findAll('.ix-collapse-panel-header')[0].trigger('click')
    await nextTick()

    expect(wrapper.findAllComponents({ name: 'IxCollapsePanel' })[0].classes()).toContain('ix-collapse-panel-active')
    expect(active.value).toEqual(['name1'])
  })

  test('accordion work', async () => {
    const active = ref('')
    const accordion = ref(true)
    const mockFn = jest.fn()

    const wrapper = mount({
      components: { IxCollapse, IxCollapsePanel },
      provide: {},
      template: `
      <ix-collapse v-model:active="active" :accordion='accordion' @change="mockFn">
        <ix-collapse-panel name="name1"> name1 </ix-collapse-panel>
        <ix-collapse-panel name="name2"> name2 </ix-collapse-panel>
        <ix-collapse-panel name="name3"> name3 </ix-collapse-panel>
      </ix-collapse>
      `,
      setup() {
        return { active, accordion, mockFn }
      },
    })

    expect(wrapper.findAll('.ix-collapse-panel').length).toBe(3)
    expect(active.value).toEqual('')
    expect(mockFn).toBeCalledTimes(0)

    active.value = 'name1'

    await nextTick()

    expect(wrapper.findAllComponents({ name: 'IxCollapsePanel' })[0].classes()).toContain('ix-collapse-panel-active')
    expect(active.value).toEqual('name1')

    await wrapper.findAll('.ix-collapse-panel-header')[1].trigger('click')

    expect(mockFn).toBeCalledTimes(1)
    expect(active.value).toEqual(['name2'])

    await wrapper.findAll('.ix-collapse-panel-header')[2].trigger('click')

    expect(mockFn).toBeCalledTimes(2)
    expect(active.value).toEqual(['name3'])

    await wrapper.findAll('.ix-collapse-panel-header')[2].trigger('click')

    expect(mockFn).toBeCalledTimes(3)
  })

  test('borderless work', async () => {
    const borderless = ref(true)
    const active = ref([])
    const wrapper = mount({
      components: { IxCollapse, IxCollapsePanel },
      template: `
      <ix-collapse v-model:active="active" :borderless='borderless'>
        <ix-collapse-panel name="name1"> name1 </ix-collapse-panel>
        <ix-collapse-panel name="name2"> name2 </ix-collapse-panel>
        <ix-collapse-panel name="name3"> name3 </ix-collapse-panel>
      </ix-collapse>
      `,
      setup() {
        return { active, borderless }
      },
    })

    expect(wrapper.findAllComponents({ name: 'IxCollapse' })[0].classes()).toContain('ix-collapse-borderless')

    borderless.value = false

    await nextTick()

    expect(wrapper.findAllComponents({ name: 'IxCollapse' })[0].classes()).not.toContain('ix-collapse-borderless')
  })

  test('collapse panel disabled work', async () => {
    const active = ref([]) as Ref<string[]>
    const disabled = ref(true)
    const mockFn = jest.fn()

    const wrapper = mount({
      components: { IxCollapse, IxCollapsePanel },
      template: `
      <ix-collapse v-model:active="active" @change="mockFn">
        <ix-collapse-panel name="name1" :disabled='disabled'> name1 </ix-collapse-panel>
        <ix-collapse-panel name="name2"> name2 </ix-collapse-panel>
        <ix-collapse-panel name="name3"> name3 </ix-collapse-panel>
      </ix-collapse>
      `,
      setup() {
        return { active, disabled, mockFn }
      },
    })

    expect(wrapper.findAllComponents({ name: 'IxCollapsePanel' })[0].classes()).toContain('ix-collapse-panel-disabled')
    expect(mockFn).toBeCalledTimes(0)

    await wrapper.findAll('.ix-collapse-panel-header')[0].trigger('click')

    expect(mockFn).toBeCalledTimes(0)

    disabled.value = false

    await nextTick()

    expect(wrapper.findAllComponents({ name: 'IxCollapsePanel' })[0].classes()).not.toContain(
      'ix-collapse-panel-disabled',
    )

    await wrapper.findAll('.ix-collapse-panel-header')[0].trigger('click')

    expect(mockFn).toBeCalledTimes(1)
  })

  test('collapse panel icon work', async () => {
    const active = ref([]) as Ref<string[]>
    const icon = ref([]) as Ref<string[]>

    const wrapper = mount({
      components: { IxCollapse, IxCollapsePanel },
      template: `
      <ix-collapse v-model:active="active">
        <ix-collapse-panel name="name1" :icon=icon> name1 </ix-collapse-panel>
        <ix-collapse-panel name="name2" :icon=icon> name2 </ix-collapse-panel>
        <ix-collapse-panel name="name3" :icon=icon> name3 </ix-collapse-panel>
      </ix-collapse>
      `,
      setup() {
        return { active, icon }
      },
    })

    expect(wrapper.findAll('.ix-icon')[0].classes()).toContain('ix-icon')

    icon.value = ['caret-right', 'caret-down']

    await nextTick()

    expect(wrapper.findAll('.ix-icon')[0].classes()).toContain('ix-icon-caret-right')
  })
})
