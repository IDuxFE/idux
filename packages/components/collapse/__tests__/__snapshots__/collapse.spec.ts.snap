// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collapse panel work header work 1`] = `
"<div class=\\"ix-collapse\\">
  <div class=\\"ix-collapse-panel\\">
    <div class=\\"ix-header ix-header-md\\">
      <div class=\\"ix-header-content\\"><span class=\\"ix-header-prefix\\"><i class=\\"ix-icon ix-icon-right\\" style=\\"transform: rotate(0deg);\\" role=\\"img\\" aria-label=\\"right\\"></i></span>
        <!----><span class=\\"ix-header-title\\">header 0</span>
        <!---->
        <!---->
      </div>
      <!---->
    </div>
    <transition-stub>
      <div class=\\"ix-collapse-panel-content\\" style=\\"display: none;\\">
        <div class=\\"ix-collapse-panel-content-box\\">
          <!---->
        </div>
      </div>
    </transition-stub>
  </div>
</div>"
`;

exports[`Collapse panel work header work 2`] = `
"<div class=\\"ix-collapse\\">
  <div class=\\"ix-collapse-panel\\">
    <div class=\\"ix-header ix-header-md\\">
      <div class=\\"ix-header-content\\"><span class=\\"ix-header-prefix\\"><i class=\\"ix-icon ix-icon-right\\" style=\\"transform: rotate(0deg);\\" role=\\"img\\" aria-label=\\"right\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" aria-hidden=\\"true\\" data-icon=\\"right\\"><path d=\\"m728.896 521.024-388.48 388.544a12.8 12.8 0 0 1-18.176 0l-27.136-27.136a12.8 12.8 0 0 1 0-18.112L638.4 521.024a12.8 12.8 0 0 0 0-18.048L295.04 159.68a12.8 12.8 0 0 1 0-18.112l27.136-27.136a12.8 12.8 0 0 1 18.112 0l388.48 388.48a12.8 12.8 0 0 1 0 18.112z\\"></path></svg></i></span>
        <!----><span class=\\"ix-header-title\\">hello header</span>
        <!----><span class=\\"ix-header-suffix\\"><i class=\\"ix-icon ix-icon-left\\" role=\\"img\\" aria-label=\\"left\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" aria-hidden=\\"true\\" data-icon=\\"left\\"><path d=\\"m295.104 502.976 388.48-388.544a12.8 12.8 0 0 1 18.176 0l27.136 27.136a12.8 12.8 0 0 1 0 18.112L385.6 502.976a12.8 12.8 0 0 0 0 18.048L728.896 864.32a12.8 12.8 0 0 1 0 18.112l-27.136 27.136a12.8 12.8 0 0 1-18.112 0l-388.48-388.48a12.8 12.8 0 0 1 0-18.112z\\"></path></svg></i></span>
      </div>
      <!---->
    </div>
    <transition-stub>
      <div class=\\"ix-collapse-panel-content\\" style=\\"display: none;\\">
        <div class=\\"ix-collapse-panel-content-box\\">
          <!---->
        </div>
      </div>
    </transition-stub>
  </div>
</div>"
`;

exports[`Collapse render work 1`] = `
"<div class=\\"ix-collapse\\">
  <div class=\\"ix-collapse-panel\\">
    <div class=\\"ix-header ix-header-md\\">
      <div class=\\"ix-header-content\\"><span class=\\"ix-header-prefix\\"><i class=\\"ix-icon ix-icon-right\\" style=\\"transform: rotate(0deg);\\" role=\\"img\\" aria-label=\\"right\\"></i></span>
        <!----><span class=\\"ix-header-title\\">header 0</span>
        <!---->
        <!---->
      </div>
      <!---->
    </div>
    <transition-stub>
      <div class=\\"ix-collapse-panel-content\\" style=\\"display: none;\\">
        <div class=\\"ix-collapse-panel-content-box\\">
          <!---->
        </div>
      </div>
    </transition-stub>
  </div>
  <div class=\\"ix-collapse-panel ix-collapse-panel-expanded\\">
    <div class=\\"ix-header ix-header-md\\">
      <div class=\\"ix-header-content\\"><span class=\\"ix-header-prefix\\"><i class=\\"ix-icon ix-icon-right\\" style=\\"transform: rotate(90deg);\\" role=\\"img\\" aria-label=\\"right\\"></i></span>
        <!----><span class=\\"ix-header-title\\">header 1</span>
        <!---->
        <!---->
      </div>
      <!---->
    </div>
    <transition-stub>
      <div class=\\"ix-collapse-panel-content\\">
        <div class=\\"ix-collapse-panel-content-box\\">
          <!---->
        </div>
      </div>
    </transition-stub>
  </div>
  <div class=\\"ix-collapse-panel\\">
    <div class=\\"ix-header ix-header-md\\">
      <div class=\\"ix-header-content\\"><span class=\\"ix-header-prefix\\"><i class=\\"ix-icon ix-icon-right\\" style=\\"transform: rotate(0deg);\\" role=\\"img\\" aria-label=\\"right\\"></i></span>
        <!----><span class=\\"ix-header-title\\">header 2</span>
        <!---->
        <!---->
      </div>
      <!---->
    </div>
    <transition-stub>
      <div class=\\"ix-collapse-panel-content\\" style=\\"display: none;\\">
        <div class=\\"ix-collapse-panel-content-box\\">
          <!---->
        </div>
      </div>
    </transition-stub>
  </div>
</div>"
`;
